src/editor/codemirror/edits.ts

```diff
@@ -59,145 +59,132 @@ class AliasesNotSupportedError extends Error {}
     // Two blank lines separating the imports from everything else.
-    changes[changes.length - 1].insert += "\n\n";
+    importChanges[importChanges.length - 1].insert += "\n\n";
   }
-  let importLines = changes
+  const importLines = importChanges
     .map((c) => c.insert.split("\n").length - 1)
     .reduce((acc, cur) => acc + cur, 0);
 
-  let additionInsertPoint: number = -1;
-  let additionPrefix = "";
-  let indent;
-  if (addition) {
+  let mainFrom: number = -1;
+  let mainPreceedingWhitespace = "";
+  let mainChange: SimpleChangeSpec | undefined;
+  let mainIndent = "";
+  if (mainCode) {
     if (line !== undefined) {
       // Tweak so the addition preview is under the mouse even if we added imports.
       line = Math.max(1, line - importLines);
       const extraLines = line - state.doc.lines;
       if (extraLines > 0) {
-        additionInsertPoint = state.doc.length;
-        additionPrefix = "\n".repeat(extraLines);
+        mainFrom = state.doc.length;
+        mainPreceedingWhitespace = "\n".repeat(extraLines);
       } else {
-        additionInsertPoint = state.doc.line(line).from;
+        mainFrom = state.doc.line(line).from;
       }
     } else {
       // When no line is specified, insert before the code (not just after the imports).
-      additionInsertPoint = skipWhitespaceLines(
-        state.doc,
-        importInsertPoint.from
-      );
+      mainFrom = skipWhitespaceLines(state.doc, importInsertPoint.from);
     }
 
-    const insertLine = state.doc.lineAt(additionInsertPoint);
-    indent = insertLine.text.match(/^(\s*)/);
-    changes.push({
-      from: additionInsertPoint,
-      insert:
-        additionPrefix + indentBy(addition, indent ? indent[0] : "") + "\n",
-    });
+    const insertLine = state.doc.lineAt(mainFrom);
+    mainIndent = insertLine.text.match(/^(\s*)/)?.[0] ?? "";
+    mainChange = {
+      from: mainFrom,
+      insert: mainPreceedingWhitespace + indentBy(mainCode, mainIndent) + "\n",
+    };
   }
 
   if (importInsertPoint.used) {
-    changes[0].insert = importInsertPoint.prefix + changes[0].insert;
+    const firstChange = importChanges[0] ?? mainChange;
+    firstChange.insert = importInsertPoint.prefix + firstChange.insert;
   }
+  const selection = calculateNewSelection(
+    mainCode,
+    type,
+    importChanges,
+    mainChange,
+    mainPreceedingWhitespace,
+    mainIndent
+  );
+  return state.update({
+    userEvent: `dnd.drop.${type}`,
+    changes: [...importChanges, ...(mainChange ? [mainChange] : [])],
+    scrollIntoView: true,
+    selection,
+  });
+};
 
+const calculateNewSelection = (
+  mainCode: string,
+  type: CodeInsertType,
+  importChanges: SimpleChangeSpec[],
+  mainChange: SimpleChangeSpec | undefined,
+  mainPreceedingWhitespace: string,
+  mainIndent: string
+): { anchor: number } | undefined => {
+  if (!mainChange) {
+    return undefined;
+  }
+  const importLength = importChanges.flatMap((c) => c.insert).join("").length;
```

Comment:
Raised https://github.com/microbit-foundation/python-editor-next/issues/424 to think about the case where this causes the wrong selection, as really the major issue is the broken code not the offset selection.